### PR TITLE
Update ESMF from JCSDA/NOAA-EMC spack fork (esmf@8.3.0 with external parallelio)

### LINKED ISSUE #31122: ESMF 8.3.0 released with changes regarding internal/external PIO
### Summary

ESMF 8.3.0 was released a few days ago. This release can now use an external parallelio (pio) installation, as long as the version is 2.5.7 at least.

Alternatively, the previous options to build ESMF without pio or with an internal pio still exist. The latter now uses a more recent version of pio (2.x instead of 1.x), which breaks the current package in spack, because it requires cmake and a few additional compiler settings.

My recommendation is for ESMF 8.3.0+ to only allow for two options: (1) turn off pio and (2) use external pio built in spack-stack. The previous approach of building its own pio while also being able to build another pio in spack-stack was not consistent.

I will be working on this update shortly.

### Rationale

We want a consistent build of ESMF with one and one one version of parallelio in the spack software stack.

### Description

_No response_

### Additional information

_No response_

### General information

- [X] I have run `spack --version` and reported the version of Spack
- [X] I have searched the issues of this repo and believe this is not a duplicate

### PR DESCRIPTION
**Pull request by climbfuji**
Update ESMF package.py from the JCSDA/NOAA-EMC spack fork. This provides updates for ESMF 8.3.0, which uses an external parallelio.

Fixes https://github.com/spack/spack/issues/31122
Replaces https://github.com/spack/spack/pull/32038

### CODE REVIEW

**Comment by climbfuji** [comment]
@jedwards4b FYI

File: var/spack/repos/builtin/packages/esmf/package.py [hunk 1]
```diff
@@ -17,21 +17,59 @@ class Esmf(MakefilePackage):
...
-    version("8.1.1", sha256="58c2e739356f21a1b32673aa17a713d3c4af9d45d572f4ba9168c357d586dc75")
-    version("8.0.1", sha256="9172fb73f3fe95c8188d889ee72fdadb4f978b1d969e1d8e401e8d106def1d84")
-    version("8.0.0", sha256="051dca45f9803d7e415c0ea146df15ce487fb55f0fce18ca61d96d4dba0c8774")
-    version("7.1.0r", sha256="ae9a5edb8d40ae97a35cbd4bd00b77061f995c77c43d36334dbb95c18b00a889")
+    # Develop is a special name for spack and is always considered the newest version
+    version("develop", branch="develop")
+    version(
+        "8.3.0",
+        sha256="0ff43ede83d1ac6beabd3d5e2a646f7574174b28a48d1b9f2c318a054ba268fd",
+    )
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/esmf/package.py:29]: Confirmed sha256 and existence of commit for new version.

However, the sha256 confirmation was not through the usual `spack checksum esmf 8.3.0`.  See comments for `url_for_version` below for fixing the URL so it works.

File: var/spack/repos/builtin/packages/esmf/package.py [hunk 2]
```diff
@@ -79,10 +118,13 @@ def url_for_version(self, version):
...
                 version.underscored
             )
-        else:
+        elif version < Version("8.2.1"):
             return "https://github.com/esmf-org/esmf/archive/ESMF_{0}.tar.gz".format(
                 version.underscored
             )
+        else:
+            # Starting with ESMF 8.2.1 releases are now in the form vx.y.z
+            return "[https://github.com/esmf-org/esmf/archive/v{0}.tar.gz]".format(version)
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/esmf/package.py]:
```suggestion
            return "https://github.com/esmf-org/esmf/archive/refs/tags/v{0}.tar.gz".format(version.dotted)
```

**Review by tldahlgren — CHANGES_REQUESTED** [review]
There are some issues with the URL for the new version.

File: var/spack/repos/builtin/packages/esmf/package.py (hunk 2 above)
Comment (climbfuji) [reply var/spack/repos/builtin/packages/esmf/package.py]: @tldahlgren Thanks for that upate.

**Comment by tldahlgren** [comment]
@spackbot fix style

**Review by tldahlgren — APPROVED** [review]
LGTM

**Pull request merged**